src/arch/x86/x86_64/src/bzimage.rs

```diff
@@ -0,0 +1,352 @@
+        // memory from 64K to LOW_MEM_1M is reserved
+        let entry_reserved = E820Entry { addr: LOW_MEM_64K as u64, size: (LOW_MEM_1M - LOW_MEM_64K) as u64, r#type: E820::RESERVED as u32 };
+        // LOW_MEM_1M to low_mem_size for ramdisk and multiboot
+        let entry_low_main = E820Entry { addr: LOW_MEM_1M as u64, size: (self.low_mem_size - LOW_MEM_1M as u64), r#type: E820::RAM as u32 };
+        // Memory between low_mem_size and high_mem is used for PCI address space
+        // main memory above 4GB
+        let entry_main = E820Entry { addr: self.high_mem_start as u64, size: self.high_mem_size as u64, r#type: E820::RAM as u32 };
+        bp.e820_table[0] = entry_first_page;
+        bp.e820_table[1] = entry_low;
+        bp.e820_table[2] = entry_reserved;
+        bp.e820_table[3] = entry_low_main;
+        bp.e820_table[4] = entry_main;
+        bp.e820_entries = 5;
+    }
+
+    /// Run the payload. This might not return.
+    pub fn run(&self, w: &mut print::WriteTo) {
+        // Jump to the payload.
+        // See: linux/Documentation/arm/Booting
+        unsafe {
+            let f = transmute::<usize, EntryPoint>(self.entry);
+            write!(w, "on to {:#x} ", self.entry).unwrap();
+            f(1, DTB_ADDR);
+        }
+        // TODO: error when payload returns.
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn setup_e820_tables_correctly_sets_table() {
+        let test_image = BzImage { low_mem_size: 0xFFFFFF, high_mem_start: 0xFFFF, high_mem_size: 0, rom_base: 0xFF_000_000, rom_size: 0xAAAA, load: 0xAAAA, entry: 0xAAAA };
+        let mut test_bp = BootParams::default();
+        let test_hdr = SetupHeader {
+            setup_sects: 26,
+            root_flags: 1,
+            syssize: 40892,
+            ram_size: 0,
+            vid_mode: 65535,
+            root_dev: 0,
+            boot_flag: 43605,
+            jump: 27371,
+            header: 1400005704,
+            version: 527,
+            realmode_swtch: 0,
+            start_sys: 4096,
+            kernel_version: 12736,
+            type_of_loader: 0,
+            loadflags: 1,
+            setup_move_size: 32768,
+            code32_start: 1048576,
+            ramdisk_image: 0,
+            ramdisk_size: 0,
+            bootsect_kludge: 0,
+            heap_end_ptr: 19056,
+            ext_loader_ver: 0,
+            ext_loader_type: 0,
+            cmd_line_ptr: 0,
+            initrd_addr_max: 2147483647,
+            kernel_alignment: 2097152,
+            relocatable_kernel: 0,
+            min_alignment: 21,
+            xloadflags: 97,
+            cmdline_size: 2047,
+            hardware_subarch: 0,
+            hardware_subarch_data: 0,
+            payload_offset: 740,
+            payload_length: 633540,
+            setup_data: 0,
+            pref_address: 16777216,
+            init_size: 11591680,
+            handover_offset: 0,
+            kernel_info_offset: 652280,
```

Comment:
Since you've implemented Default, you should be able to do `...Default::default()`.

Reply:
That's a lot neater.